### ISSUE TITLE
Enabling G1 garbage collector causes compilation error

### ISSUE THREAD

**Issue by apetrelli** [oracle/graal#3124]
When I compile my project using GraalVM EE 20.3 using Maven and its native-image plugin, I receive the following error:

```
Fatal error:com.oracle.svm.core.util.VMError$HostedError: Could not find field g.alloc
        at com.oracle.svm.core.util.VMError.shouldNotReachHere(VMError.java:68)
        at com.oracle.svm.enterprise.gc.g1.hosted.a.a(stripped:185)
        at com.oracle.svm.enterprise.gc.g1.hosted.a.b(stripped:152)
        at com.oracle.svm.enterprise.gc.g1.hosted.b.beforeAnalysis(stripped:129)
        at com.oracle.svm.hosted.NativeImageGenerator.lambda$runPointsToAnalysis$7(NativeImageGenerator.java:696)
        at com.oracle.svm.hosted.FeatureHandler.forEachFeature(FeatureHandler.java:70)
        at com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:696)
        at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:558)
        at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$0(NativeImageGenerator.java:471)
        at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1407)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Error: Image build request failed with exit status 1
com.oracle.svm.driver.NativeImage$NativeImageError: Image build request failed with exit status 1
        at com.oracle.svm.driver.NativeImage.showError(NativeImage.java:1647)
        at com.oracle.svm.driver.NativeImage.build(NativeImage.java:1397)
        at com.oracle.svm.driver.NativeImage.performBuild(NativeImage.java:1358)
        at com.oracle.svm.driver.NativeImage.main(NativeImage.java:1317)
        at com.oracle.svm.driver.NativeImage$JDK9Plus.main(NativeImage.java:1829)
```
I used this project to test this problem:
https://github.com/apetrelli/scafa
with branch "web-server-native", module scafa-web-server-linux.
Compiling without G1 works fine.

**Describe GraalVM and your environment:**
 - GraalVM version : 20.3 EE
 -  - JDK major version: 1
 - OS: LInux, Kubuntu 20.04
 - Architecture: AMD64

**More details**
This is the complete result from the compilation.

```
[INFO] Executing: /home/antonio/javadev/libs/graalvm-ee/bin/native-image -cp /home/antonio/.m2/repository/com/github/apetrelli/scafa/scafa-web-server/1.0.0-SNAPSHOT/scafa-web-server-1.0.0-SNAPSHOT.jar:/home/antonio/.m2/repository/com/github/apetrelli/scafa/scafa-http-server/1.0.0-SNAPSHOT/scafa-http-server-1.0.0-SNAPSHOT.jar:/home/antonio/.m2/repository/com/github/apetrelli/scafa/scafa-http/1.0.0-SNAPSHOT/scafa-http-1.0.0-SNAPSHOT.jar:/home/antonio/.m2/repository/com/github/apetrelli/scafa/scafa-proto/1.0.0-SNAPSHOT/scafa-proto-1.0.0-SNAPSHOT.jar:/home/antonio/.m2/repository/com/github/apetrelli/scafa/scafa-tls/1.0.0-SNAPSHOT/scafa-tls-1.0.0-SNAPSHOT.jar:/home/antonio/.m2/repository/com/github/apetrelli/scafa/scafa-http-gateway/1.0.0-SNAPSHOT/scafa-http-gateway-1.0.0-SNAPSHOT.jar:/home/antonio/.m2/repository/org/ini4j/ini4j/0.5.4/ini4j-0.5.4.jar:/home/antonio/git/scafa/scafa-web-server-linux/target/scafa-web-server-linux-1.0.0-SNAPSHOT.jar --verbose --no-server -H:EnableURLProtocols=http -H:+AllowVMInspection --gc=G1 --native-image-info --verbose -H:Class=com.github.apetrelli.scafa.web.SyncScafaWebServerMain -H:Name=scafa
Executing [
/home/antonio/javadev/libs/graalvm-ee/bin/java \
-XX:+UseParallelGC \
-XX:+UnlockExperimentalVMOptions \
-XX:+EnableJVMCI \
-Dtruffle.TrustAllTruffleRuntimeProviders=true \
-Dtruffle.TruffleRuntime=com.oracle.truffle.api.impl.DefaultTruffleRuntime \
-Dgraalvm.ForcePolyglotInvalid=true \
-Dgraalvm.locatorDisabled=true \
-Dsubstratevm.IgnoreGraalVersionCheck=true \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.aarch64=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.amd64=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.code.site=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.code.stack=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.code=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.common=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.hotspot.aarch64=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.hotspot.amd64=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.hotspot.sparc=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.hotspot=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.meta=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.runtime=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.services=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.sparc=ALL-UNNAMED \
--add-exports=org.graalvm.truffle/com.oracle.truffle.api=ALL-UNNAMED \
--add-opens=jdk.internal.vm.compiler/org.graalvm.compiler.debug=ALL-UNNAMED \
--add-opens=jdk.internal.vm.compiler/org.graalvm.compiler.nodes=ALL-UNNAMED \
--add-opens=jdk.unsupported/sun.reflect=ALL-UNNAMED \
--add-opens=java.base/jdk.internal.module=ALL-UNNAMED \
--add-opens=java.base/jdk.internal.ref=ALL-UNNAMED \
--add-opens=java.base/jdk.internal.reflect=ALL-UNNAMED \
--add-opens=java.base/java.io=ALL-UNNAMED \
--add-opens=java.base/java.lang=ALL-UNNAMED \
--add-opens=java.base/java.lang.reflect=ALL-UNNAMED \
--add-opens=java.base/java.lang.invoke=ALL-UNNAMED \
--add-opens=java.base/java.lang.ref=ALL-UNNAMED \
--add-opens=java.base/java.net=ALL-UNNAMED \
--add-opens=java.base/java.nio=ALL-UNNAMED \
--add-opens=java.base/java.nio.file=ALL-UNNAMED \
--add-opens=java.base/java.security=ALL-UNNAMED \
--add-opens=java.base/javax.crypto=ALL-UNNAMED \
--add-opens=java.base/java.util=ALL-UNNAMED \
--add-opens=java.base/java.util.concurrent.atomic=ALL-UNNAMED \
--add-opens=java.base/sun.security.x509=ALL-UNNAMED \
--add-opens=java.base/jdk.internal.logger=ALL-UNNAMED \
--add-opens=java.desktop/sun.java2d.cmm.profiles=ALL-UNNAMED \
--add-opens=org.graalvm.sdk/org.graalvm.nativeimage.impl=ALL-UNNAMED \
--add-opens=org.graalvm.sdk/org.graalvm.polyglot=ALL-UNNAMED \
--add-opens=org.graalvm.truffle/com.oracle.truffle.polyglot=ALL-UNNAMED \
--add-opens=org.graalvm.truffle/com.oracle.truffle.api.impl=ALL-UNNAMED \
-XX:+UseJVMCINativeLibrary \
-Xss10m \
-Xms1g \
-Xmx13353245080 \
-Duser.country=US \
-Duser.language=en \
-Djava.awt.headless=true \
-Dorg.graalvm.version=20.3.0 \
-Dorg.graalvm.config= \
-Dcom.oracle.graalvm.isaot=true \
-Djava.system.class.loader=com.oracle.svm.hosted.NativeImageSystemClassLoader \
-Xshare:off \
--module-path \
/home/antonio/javadev/libs/graalvm-ee/lib/truffle/truffle-api.jar \
-Djdk.internal.lambda.disableEagerInitialization=true \
-Djdk.internal.lambda.eagerlyInitialize=false \
-Djava.lang.invoke.InnerClassLambdaMetafactory.initializeLambdas=false \
-javaagent:/home/antonio/javadev/libs/graalvm-ee/lib/svm/builder/svm.jar \
-cp \
/home/antonio/javadev/libs/graalvm-ee/lib/svm/builder/svm-enterprise.jar:/home/antonio/javadev/libs/graalvm-ee/lib/svm/builder/javacpp-shadowed.jar:/home/antonio/javadev/libs/graalvm-ee/lib/svm/builder/pointsto.jar:/home/antonio/javadev/libs/graalvm-ee/lib/svm/builder/objectfile.jar:/home/antonio/javadev/libs/graalvm-ee/lib/svm/builder/svm-enterprise-llvm.jar:/home/antonio/javadev/libs/graalvm-ee/lib/svm/builder/llvm-wrapper-shadowed.jar:/home/antonio/javadev/libs/graalvm-ee/lib/svm/builder/svm-llvm.jar:/home/antonio/javadev/libs/graalvm-ee/lib/svm/builder/llvm-platform-specific-shadowed.jar:/home/antonio/javadev/libs/graalvm-ee/lib/svm/builder/svm.jar \
'com.oracle.svm.hosted.NativeImageGeneratorRunner$JDK9Plus' \
-watchpid \
34077 \
-imagecp \
/home/antonio/javadev/libs/graalvm-ee/lib/svm/library-support-enterprise.jar:/home/antonio/javadev/libs/graalvm-ee/lib/svm/library-support.jar:/home/antonio/git/scafa/scafa-web-server-linux/target/ext:/home/antonio/git/scafa/scafa-web-server-linux/target/lib/scafa-web-server-1.0.0-SNAPSHOT.jar:/home/antonio/git/scafa/scafa-web-server-linux/target/lib/scafa-http-server-1.0.0-SNAPSHOT.jar:/home/antonio/git/scafa/scafa-web-server-linux/target/lib/scafa-http-1.0.0-SNAPSHOT.jar:/home/antonio/git/scafa/scafa-web-server-linux/target/lib/scafa-proto-1.0.0-SNAPSHOT.jar:/home/antonio/git/scafa/scafa-web-server-linux/target/lib/scafa-tls-1.0.0-SNAPSHOT.jar:/home/antonio/git/scafa/scafa-web-server-linux/target/lib/scafa-http-gateway-1.0.0-SNAPSHOT.jar:/home/antonio/git/scafa/scafa-web-server-linux/target/lib/ini4j-0.5.4.jar:/home/antonio/.m2/repository/com/github/apetrelli/scafa/scafa-web-server/1.0.0-SNAPSHOT/scafa-web-server-1.0.0-SNAPSHOT.jar:/home/antonio/.m2/repository/com/github/apetrelli/scafa/scafa-http-server/1.0.0-SNAPSHOT/scafa-http-server-1.0.0-SNAPSHOT.jar:/home/antonio/.m2/repository/com/github/apetrelli/scafa/scafa-http/1.0.0-SNAPSHOT/scafa-http-1.0.0-SNAPSHOT.jar:/home/antonio/.m2/repository/com/github/apetrelli/scafa/scafa-proto/1.0.0-SNAPSHOT/scafa-proto-1.0.0-SNAPSHOT.jar:/home/antonio/.m2/repository/com/github/apetrelli/scafa/scafa-tls/1.0.0-SNAPSHOT/scafa-tls-1.0.0-SNAPSHOT.jar:/home/antonio/.m2/repository/com/github/apetrelli/scafa/scafa-http-gateway/1.0.0-SNAPSHOT/scafa-http-gateway-1.0.0-SNAPSHOT.jar:/home/antonio/.m2/repository/org/ini4j/ini4j/0.5.4/ini4j-0.5.4.jar:/home/antonio/git/scafa/scafa-web-server-linux/target/scafa-web-server-linux-1.0.0-SNAPSHOT.jar \
-H:Path=/home/antonio/git/scafa/scafa-web-server-linux/target \
-H:JNIConfigurationResources=META-INF/native-image/jni-config.json \
-H:ReflectionConfigurationResources=META-INF/native-image/reflect-config.json \
-H:ResourceConfigurationResources=META-INF/native-image/resource-config.json \
-H:DynamicProxyConfigurationResources=META-INF/native-image/proxy-config.json \
-H:EnableURLProtocols=http \
-H:+AllowVMInspection \
-H:+UseG1GC \
-H:+DumpTargetInfo \
-H:Class=com.github.apetrelli.scafa.web.SyncScafaWebServerMain \
-H:Name=scafa \
-H:CLibraryPath=/home/antonio/javadev/libs/graalvm-ee/lib/svm/clibraries/linux-amd64 \

]
[scafa:34096]    classlist:   2,214.10 ms,  0.96 GB
[scafa:34096]        (cap):   1,126.44 ms,  0.96 GB
[scafa:34096]        setup:   5,070.77 ms,  0.96 GB
Fatal error:com.oracle.svm.core.util.VMError$HostedError: Could not find field g.alloc
        at com.oracle.svm.core.util.VMError.shouldNotReachHere(VMError.java:68)
        at com.oracle.svm.enterprise.gc.g1.hosted.a.a(stripped:185)
        at com.oracle.svm.enterprise.gc.g1.hosted.a.b(stripped:152)
        at com.oracle.svm.enterprise.gc.g1.hosted.b.beforeAnalysis(stripped:129)
        at com.oracle.svm.hosted.NativeImageGenerator.lambda$runPointsToAnalysis$7(NativeImageGenerator.java:696)
        at com.oracle.svm.hosted.FeatureHandler.forEachFeature(FeatureHandler.java:70)
        at com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:696)
        at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:558)
        at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$0(NativeImageGenerator.java:471)
        at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1407)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Error: Image build request failed with exit status 1
com.oracle.svm.driver.NativeImage$NativeImageError: Image build request failed with exit status 1
        at com.oracle.svm.driver.NativeImage.showError(NativeImage.java:1647)
        at com.oracle.svm.driver.NativeImage.build(NativeImage.java:1397)
        at com.oracle.svm.driver.NativeImage.performBuild(NativeImage.java:1358)
        at com.oracle.svm.driver.NativeImage.main(NativeImage.java:1317)
        at com.oracle.svm.driver.NativeImage$JDK9Plus.main(NativeImage.java:1829)
```

**Comment by mcraj017** [comment]
@apetrelli thanks for reporting the issue, we will check it out and update you

**Comment by apetrelli** [comment]
I just tried with 21.0.0 version today and still I have the same problem.

**Comment by mcraj017** [comment]
@apetrelli 
I tried using mvn package to create native-image of https://github.com/apetrelli/scafa/blob/web-server-native/scafa-web-server-linux
But doesn't work
please share the steps 
you used to change https://github.com/apetrelli/scafa project to native-image

**Comment by apetrelli** [comment]
@mcraj017 
Sorry I forgot to tell that the committed version at Github the option to use G1 is not there.
Here:
https://github.com/apetrelli/scafa/blob/web-server-native/scafa-web-server-linux/pom.xml
you have to add after line 58 the following option:
--gc=G1
and you have to build everything from the main project, otherwise the dependencies are not resolved.
I tried using:
mvn install
command, but mvn package should work too. And obviously you have to use GraalVM EE to launch Maven.
If you need further information let me know.

**Comment by apetrelli** [comment]
FWIW this is the command line I use:
JAVA_HOME=~/javadev/libs/graalvm-ee mvn clean install

**Comment by mcraj017** [comment]
@apetrelli it is working with GraalVM 21.0.0.2, please check 
```
-H:CLibraryPath=/home/munish/graalvm-ee-complete-java11-21.0.0.2/lib/svm/clibraries/linux-amd64 \

]
[scafa:9469]    classlist:   1,073.46 ms,  0.96 GB
[scafa:9469]        (cap):   3,586.63 ms,  0.96 GB
[scafa:9469]        setup:   5,407.50 ms,  0.96 GB
[scafa:9469]     (clinit):     252.24 ms,  1.54 GB
[scafa:9469]   (typeflow):   7,481.08 ms,  1.54 GB
[scafa:9469]    (objects):   6,615.21 ms,  1.54 GB
[scafa:9469]   (features):     422.75 ms,  1.54 GB
[scafa:9469]     analysis:  15,157.49 ms,  1.54 GB
[scafa:9469]     universe:     691.33 ms,  1.54 GB
[scafa:9469]      (parse):     952.55 ms,  1.54 GB
[scafa:9469]     (inline):   1,900.89 ms,  1.94 GB
[scafa:9469]    (compile):  23,063.26 ms,  4.98 GB
[scafa:9469]      compile:  26,887.52 ms,  4.98 GB
[scafa:9469]        image:   1,574.86 ms,  4.98 GB
[scafa:9469]        write:   1,361.59 ms,  4.98 GB
[scafa:9469]      [total]:  52,335.71 ms,  4.98 GB
[INFO]
[INFO] --- maven-install-plugin:2.4:install (default-install) @ scafa-web-server-linux ---
[INFO] Installing /home/munish/scafa/scafa-web-server-linux/target/scafa-web-server-linux-1.0.0-SNAPSHOT.jar to /home/munish/.m2/repository/com/github/apetrelli/scafa/scafa-web-server-linux/1.0.0-SNAPSHOT/scafa-web-server-linux-1.0.0-SNAPSHOT.jar
[INFO] Installing /home/munish/scafa/scafa-web-server-linux/pom.xml to /home/munish/.m2/repository/com/github/apetrelli/scafa/scafa-web-server-linux/1.0.0-SNAPSHOT/scafa-web-server-linux-1.0.0-SNAPSHOT.pom
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  55.826 s
[INFO] Finished at: 2021-04-02T16:59:41+08:00
[INFO] ------------------------------------------------------------------------
```

**Comment by mcraj017** [comment]
@apetrelli please ignore my last comment, I forgot to add --gc=G1 flag, 
I can reproduce the issue now, I will report the bug to our dev team

**Comment by apetrelli** [comment]
> @apetrelli please ignore my last comment, I forgot to add --gc=G1 flag,
> I can reproduce the issue now, I will report the bug to our dev team

Fantastic! I would like to test G1 GC to see how it behaves in contrast to the default one, thanks!

**Comment by christianhaeubl** [comment]
@apetrelli the issue will be fixed in 21.1. As a workaround, you can remove the image build argument `-H:+AllowVMInspection`

**Comment by zakkak** [comment]
> @apetrelli the issue will be fixed in 21.1

@christianhaeubl can you please link the commit that fixes the issue as well?

**Comment by christianhaeubl** [comment]
@zakkak this issue only affects EE, so there is no publicly visible commit.
